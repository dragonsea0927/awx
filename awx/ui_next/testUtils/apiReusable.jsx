// eslint-disable-next-line import/prefer-default-export
export function describeNotificationMixin (Model, name) {
  describe(name, () => {
    const mockHttp = ({ post: jest.fn(() => Promise.resolve()) });
    const ModelAPI = new Model(mockHttp);

    afterEach(() => {
      jest.clearAllMocks();
    });

    const parameters = [
      ['success', true],
      ['success', false],
      ['error', true],
      ['error', false],
    ];
    parameters.forEach(([type, state]) => {
      const label = `[notificationType=${type}, associationState=${state}]`;
      const testName = `updateNotificationTemplateAssociation ${label} makes expected http calls`;

      test(testName, async (done) => {
        await ModelAPI.updateNotificationTemplateAssociation(1, 21, type, state);

        const expectedPath = `${ModelAPI.baseUrl}1/notification_templates_${type}/`;
        expect(mockHttp.post).toHaveBeenCalledTimes(1);

        const expectedParams = state ? { id: 21 } : { id: 21, disassociate: true };
        expect(mockHttp.post.mock.calls.pop()).toEqual([expectedPath, expectedParams]);

        done();
      });
    });
  });
}
