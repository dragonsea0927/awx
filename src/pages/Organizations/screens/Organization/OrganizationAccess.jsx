import React from 'react';

import { withNetwork } from '../../../../contexts/Network';

import OrganizationAccessList from '../../components/OrganizationAccessList';

class OrganizationAccess extends React.Component {
  constructor (props) {
    super(props);

    this.getOrgAccessList = this.getOrgAccessList.bind(this);
    this.removeRole = this.removeRole.bind(this);
  }

  getOrgAccessList (id, params) {
    const { api } = this.props;
    return api.getOrganizationAccessList(id, params);
  }

  removeRole (url, id) {
    const { api } = this.props;
    return api.disassociate(url, id);
  }

  render () {
    return (
      <OrganizationAccessList
        getAccessList={this.getOrgAccessList}
        removeRole={this.removeRole}
      />
    );
  }
}

export default withNetwork(OrganizationAccess);
