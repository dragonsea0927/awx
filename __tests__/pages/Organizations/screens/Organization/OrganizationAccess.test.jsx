import React from 'react';
import { mountWithContexts } from '../../../../enzymeHelpers';
import OrganizationAccess from '../../../../../src/pages/Organizations/screens/Organization/OrganizationAccess';
import { sleep } from '../../../../testUtils';

describe('<OrganizationAccess />', () => {
  let network;
  const organization = {
    id: 1,
    name: 'Default',
    summary_fields: {
      object_roles: {},
      user_capabilities: {
        edit: true
      }
    }
  };

  const data = {
    count: 2,
    results: [{
      id: 1,
      username: 'joe',
      url: '/foo',
      first_name: 'joe',
      last_name: 'smith',
      summary_fields: {
        direct_access: [{
          role: {
            id: 1,
            name: 'Member',
            resource_name: 'Org',
            resource_type: 'organization',
            user_capabilities: { unattach: true },
          }
        }],
        indirect_access: [],
      }
    }, {
      id: 2,
      username: 'jane',
      url: '/bar',
      first_name: 'jane',
      last_name: 'brown',
      summary_fields: {
        direct_access: [{
          role: {
            id: 3,
            name: 'Member',
            resource_name: 'Org',
            resource_type: 'organization',
            team_id: 5,
            team_name: 'The Team',
            user_capabilities: { unattach: true },
          }
        }],
        indirect_access: [],
      }
    }]
  };

  beforeEach(() => {
    network = {
      api: {
        getOrganizationAccessList: jest.fn()
          .mockReturnValue(Promise.resolve({ data })),
        disassociateTeamRole: jest.fn(),
        disassociateUserRole: jest.fn(),
        toJSON: () => '/api/',
      },
    };
  });

  test('initially renders succesfully', () => {
    const wrapper = mountWithContexts(
      <OrganizationAccess organization={organization} />,
      { context: { network } }
    );
    expect(wrapper.find('OrganizationAccess')).toMatchSnapshot();
  });

  test('should fetch and display access records on mount', async () => {
    const wrapper = mountWithContexts(
      <OrganizationAccess organization={organization} />,
      { context: { network } }
    );
    await sleep(0);
    wrapper.update();
    expect(network.api.getOrganizationAccessList).toHaveBeenCalled();
    expect(wrapper.find('OrganizationAccess').state('isInitialized')).toBe(true);
    expect(wrapper.find('PaginatedDataList').prop('items')).toEqual(data.results);
    expect(wrapper.find('OrganizationAccessItem')).toHaveLength(2);
  });

  test('should open confirmation dialog when deleting role', async () => {
    const wrapper = mountWithContexts(
      <OrganizationAccess organization={organization} />,
      { context: { network } }
    );
    await sleep(0);
    wrapper.update();

    const button = wrapper.find('ChipButton').at(0);
    button.prop('onClick')();
    wrapper.update();

    const component = wrapper.find('OrganizationAccess');
    expect(component.state('roleToDelete'))
      .toEqual(data.results[0].summary_fields.direct_access[0].role);
    expect(component.state('roleToDeleteAccessRecord'))
      .toEqual(data.results[0]);
    expect(component.find('DeleteRoleConfirmationModal')).toHaveLength(1);
  });

  it('should close dialog when cancel button clicked', async () => {
    const wrapper = mountWithContexts(
      <OrganizationAccess organization={organization} />,
      { context: { network } }
    );
    await sleep(0);
    wrapper.update();
    const button = wrapper.find('ChipButton').at(0);
    button.prop('onClick')();
    wrapper.update();

    wrapper.find('DeleteRoleConfirmationModal').prop('onCancel')();
    const component = wrapper.find('OrganizationAccess');
    expect(component.state('roleToDelete')).toBeNull();
    expect(component.state('roleToDeleteAccessRecord')).toBeNull();
    expect(network.api.disassociateTeamRole).not.toHaveBeenCalled();
    expect(network.api.disassociateUserRole).not.toHaveBeenCalled();
  });

  it('should delete user role', async () => {
    const wrapper = mountWithContexts(
      <OrganizationAccess organization={organization} />,
      { context: { network } }
    );
    await sleep(0);
    wrapper.update();
    const button = wrapper.find('ChipButton').at(0);
    button.prop('onClick')();
    wrapper.update();

    wrapper.find('DeleteRoleConfirmationModal').prop('onConfirm')();
    await sleep(0);
    wrapper.update();

    const component = wrapper.find('OrganizationAccess');
    expect(component.state('roleToDelete')).toBeNull();
    expect(component.state('roleToDeleteAccessRecord')).toBeNull();
    expect(network.api.disassociateTeamRole).not.toHaveBeenCalled();
    expect(network.api.disassociateUserRole).toHaveBeenCalledWith(1, 1);
    expect(network.api.getOrganizationAccessList).toHaveBeenCalledTimes(2);
  });

  it('should delete team role', async () => {
    const wrapper = mountWithContexts(
      <OrganizationAccess organization={organization} />,
      { context: { network } }
    );
    await sleep(0);
    wrapper.update();
    const button = wrapper.find('ChipButton').at(1);
    button.prop('onClick')();
    wrapper.update();

    wrapper.find('DeleteRoleConfirmationModal').prop('onConfirm')();
    await sleep(0);
    wrapper.update();

    const component = wrapper.find('OrganizationAccess');
    expect(component.state('roleToDelete')).toBeNull();
    expect(component.state('roleToDeleteAccessRecord')).toBeNull();
    expect(network.api.disassociateTeamRole).toHaveBeenCalledWith(5, 3);
    expect(network.api.disassociateUserRole).not.toHaveBeenCalled();
    expect(network.api.getOrganizationAccessList).toHaveBeenCalledTimes(2);
  });
});
