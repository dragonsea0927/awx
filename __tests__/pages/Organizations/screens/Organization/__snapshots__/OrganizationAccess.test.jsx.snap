// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrganizationAccess /> initially renders succesfully 1`] = `
<OrganizationAccess
  handleHttpError={[Function]}
  history={"/history/"}
  i18n={"/i18n/"}
  location={
    Object {
      "hash": "",
      "pathname": "",
      "search": "",
      "state": "",
    }
  }
  match={
    Object {
      "isExact": false,
      "params": Object {},
      "path": "",
      "url": "",
    }
  }
  organization={
    Object {
      "id": 1,
      "name": "Default",
      "summary_fields": Object {
        "object_roles": Object {},
        "user_capabilities": Object {
          "edit": true,
        },
      },
    }
  }
>
  <div>
    Loading...
  </div>
</OrganizationAccess>
`;
